controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1424,6 +1437,192 @@ func (r *BareMetalHostReconciler) manageHostPower(prov provisioner.Provisioner,
 }
 
+// DataImage handler for attaching/detaching image.
+func (r *BareMetalHostReconciler) handleDataImageActions(prov provisioner.Provisioner, info *reconcileInfo) actionResult {
+	dataImage := &metal3api.DataImage{}
+	if err := r.Get(info.ctx, info.request.NamespacedName, dataImage); err != nil {
+		// DataImage does not exist or it may have been deleted
+		if k8serrors.IsNotFound(err) {
+			info.log.Info("dataImage not found")
+			return nil
+		}
+		// Error reading the object - requeue the request.
+		return actionError{fmt.Errorf("could not load dataImage, %w", err)}
+	}
+
+	// Set ControllerReference to DataImage
+	if !ownerReferenceExists(info.host, dataImage) {
+		if err := controllerutil.SetControllerReference(info.host, dataImage, r.Scheme()); err != nil {
+			return actionError{fmt.Errorf("could not set bmh as controller, %w", err)}
+		}
+		if err := r.Update(info.ctx, dataImage); err != nil {
+			return actionError{fmt.Errorf("failure updating dataImage status, %w", err)}
+		}
+
+		return actionContinue{}
+	}
+
+	// Check if any attach/detach action is pending or failed to attach
+	isNodeBusy, nodeError := prov.IsDataImageReady()
+	if isNodeBusy {
+		info.log.Info("Node is busy, requeuing")
+
+		if nodeError != nil {
+			return actionError{nodeError}
+		}
+
+		return actionContinue{dataImageRetryDelay}
+	}
+
+	// Is the current dataImage status valid
+	dirty := false
+	// In case the last node error was not nil for dataimage,
+	// update message and counter
+	if nodeError != nil {
+		dataImage.Status.Error.Message = nodeError.Error()
+		dataImage.Status.Error.Count++
+		dirty = true
+	}
+
+	deleteDataImage := !dataImage.DeletionTimestamp.IsZero()
+
+	requestedURL := dataImage.Spec.URL
+
+	// We can assume non null value since we initialized the status earlier
+	attachedURL := dataImage.Status.AttachedImage.URL
+
+	if deleteDataImage {
+		info.log.Info("DataImage requested for deletion")
+		if attachedURL != "" || dirty {
+			info.log.Info("Detaching DataImage as it was deleted")
+			err := r.detachDataImage(prov, info, dataImage)
+			if err != nil {
+				return actionError{fmt.Errorf("failed to detach, %w", err)}
+			}
+
+			// Requeue to give time to the DataImage Reconciler to update the
+			// status. In case of failure, we will enter this section and
+			// detachDataImage will be called again -> can this cause issues ?
+			return actionContinue{dataImageUpdateDelay}
+		}
+
+		return nil
+	}
+
+	if requestedURL != attachedURL || dirty {
+		info.log.Info("DataImage change detected")
+		if attachedURL != "" {
+			info.log.Info("Detaching DataImage")
+			err := r.detachDataImage(prov, info, dataImage)
+			if err != nil {
+				return actionError{fmt.Errorf("failed to detach, %w", err)}
+			}
+
+			// Requeue to give time to the DataImage Reconciler to update the
+			// status. In case of failure, we will enter this section and
+			// detachDataImage will be called again -> can this cause issues ?
```

Comment:
Is this question resolved?

Reply:
I see the logic now: we first set the attached image in the status, then simply wait for the reservation to be removed without an error. Works for me.